### ISSUE TITLE
Lack of assertion of mobile platform robust isolation of app access to keychain

### ISSUE THREAD

**Issue by ovitrif** [synonymdev/bitkit#1654]
## `p5` React-native-keychain

- Lack of usage of the accessibility, access groups settings for both platforms
  - LOSS OF FUNDS risk
- Lack of usage of highest level of secure storage on the Android platform
  - LOSS OF FUNDS  risk
  - Unsafe secret isolation among a set of applications running on a user mobile phone.
    - `p7` Custom `KeychainModuleBuilder` not used to leverage the advanced Android KeyStore API
      > I suspect it isn't correct, will check in detail to confirm.

### `p8-9` Review of iOS support

- Library relies on a number of deprecated iOS functions
  - We aren't using any of these.
- `p9` Library using 3rd least restrictive `accessibility` flag as default (ie. when not specified)
- Secrets stored in a dynamically allocated buffer when fetched from keychain
  - This API function does not offer apparent zeroing guarantees of the used memory buffer once it is deallocated.
- Ambiguous error management of secret resetting apis
  - Contains error silencing which can be exploited by a malicious app wiping out a target secret storage, then preventing the legitimate calling thread in the honest app to detect tampering with its secret storage.

### `p9-p12` Review of Android support

- SharedPreferences used to store secrets
  - This mechanism does not provide auth at entry removed, which could lead to a stale state, where an app evaluates entries to be removed but they're still present.
  > We switched to DataStore via our own fork of the library.
- Plausible downgrade issue with the decrypt credential helper used in `getGenericPasswordForOptions` which Bitkit leverages.
  - `decryptCredential` method evaluates if encrypted storage level-up should be performed as the logged storage is detected not equal to the current set of options passed (L645 in `KeychainModule.java`).
    - The necessity of upgrading is gated on rules values that can be themselves provided as called-selected options in `getGenericPassForOptions`. The code can be more strict and store available storages by level of strength.
    - There is no available unit test for `decryptCredential` to invalidate this observation.
- Lack of application-specific password provided to extract/remove key
  - A malicious app could call `extractKey()` with a deduced `alias` argument, gaining cleartext access to username / password.
  - No authenticated access used by `deleteEntry` and `removeKey`.

  ### `p12+` Library API and Bitkit usage

- BitKit uses default Keychain builder
  - At `setGenericPassword()` call, there is no definition of a set of options settings `accessControl`, `accessGroup` or `accessible`.
    - In the lack of such options, a 3rd-party app can access Bitkit's secrets by leveraging iOS Keystore methods (ie. `SecItemCopyMatching()`) or Java Keystore methods (ie. `getKey()`) from their local application contexts.
- Private key extraction risk
- A malicious 3rd-party app can launch few other types of attacks e.g a replay attack where the honest app mnemonic is deleted from local keystore then replaced by a new mnemonic known by the attacker, where all `bip32` derived payments keys can be then silently snitched.
- A wipe out attack can happen where the honest app mnemonic is extracted to an external server then the correct secret state is wiped out from the mobile device, in an attempt to blackmail the user from its funds.

These attacks can lead to user LOSS OF FUNDS risks.

- Recommended to set application-specific passwords (i.e `AccessControl.APPLICATION_PASSWORD`) to ensure safe segregation of Bitkit iOS and Android keystore accesses.

    The necessity of upgrading is gated on rules values that can be themselves provided as called-selected options in `getGenericPassForOptions`. The code can be more strict and store available storages by level of strength. There is no available unit test code about `decryptCredential` to invalidate this observation.

- Lack of application-specific password provided to extract/remove key

  - A malicious app could call `extractKey()` with a deduced `alias` argument, gaining cleartext access to username / password.
  - No authenticated access used by `deleteEntry` and `removeKey`.

### `p12+` Library API and Bitkit usage

- BitKit uses default Keychain builder
  -  Custom `KeychainModuleBuilder` not used to leverage the advanced Android KeyStore API
      > I suspect this is  incorrect, will check in detail to confirm.
  - At `setGenericPassword()` call, there is no definition of a set of options settings `accessControl`, `accessGroup` or `accessible`.
    - In the lack of such options, a 3rd-party app can access Bitkit's secrets by leveraging iOS Keystore methods (ie. `SecItemCopyMatching()`) or Java Keystore methods (ie. `getKey()`) from their local application contexts.
- Private key extraction risk
- A malicious 3rd-party app can launch few other types of attacks e.g a replay attack where the honest app mnemonic is deleted from local keystore then replaced by a new mnemonic known by the attacker, where all `bip32` derived payments keys can be then silently snitched.
- A wipe out attack can happen where the honest app mnemonic is extracted to an external server then the correct secret state is wiped out from the mobile device, in an attempt to blackmail the user from its funds.

Those attacks can lead to user LOSS OF FUNDS risks.

- Recommend setting application-specific passwords (i.e `AccessControl.APPLICATION_PASSWORD`) to ensure safe segregation of Bitkit iOS and Android keystore accesses.
